synapse/storage/databases/main/registration.py

```diff
@@ -1760,6 +1760,96 @@ def start_or_continue_validation_session_txn(txn):
             start_or_continue_validation_session_txn,
         )
 
+    async def registration_token_is_valid(self, token: str) -> bool:
+        """Checks if a token can be used to authenticate a registration.
+
+        Args:
+            token: The registration token to be checked
+        """
+        res = await self.db_pool.simple_select_one(
+            "registration_tokens",
+            keyvalues={"token": token},
+            retcols=["uses_allowed", "pending", "completed", "expiry_time"],
+            allow_none=True,
+        )
+
+        # Check if the token exists
+        if res is None:
+            return False
+
+        # Check if the token has expired
+        now = self._clock.time_msec()
+        if res["expiry_time"] and res["expiry_time"] < now:
+            return False
+
+        # Check if the token has been used up
+        if (
+            res["uses_allowed"]
+            and res["pending"] + res["completed"] >= res["uses_allowed"]
+        ):
+            return False
+
+        # Otherwise, the token is valid
+        return True
+
+    async def set_registration_token_pending(self, token: str) -> None:
+        """Increment the pending registrations counter for a token.
+
+        Args:
+            token: The registration token pending use
+        """
+
+        def _set_registration_token_pending_txn(txn):
+            pending = self.db_pool.simple_select_one_onecol_txn(
+                txn,
+                "registration_tokens",
+                keyvalues={"token": token},
+                retcol="pending",
+            )
+            self.db_pool.simple_update_one_txn(
+                txn,
+                "registration_tokens",
+                keyvalues={"token": token},
+                updatevalues={"pending": pending + 1},
+            )
+
+        return await self.db_pool.runInteraction(
+            "set_registration_token_pending", _set_registration_token_pending_txn
+        )
+
+    async def use_registration_token(self, token: str) -> None:
+        """Complete a use of the given registration token.
+
+        The `pending` counter will be decremented, and the `completed`
+        counter will be incremented.
+
+        Args:
+            token: The registration token to be 'used'
+        """
+
+        def _use_registration_token_txn(txn):
+            res = self.db_pool.simple_select_one_txn(
+                txn,
+                "registration_tokens",
+                keyvalues={"token": token},
+                retcols=["pending", "completed"],
+            )
+
+            # Decrement pending and increment completed
+            self.db_pool.simple_update_one_txn(
+                txn,
+                "registration_tokens",
+                keyvalues={"token": token},
+                updatevalues={
+                    "completed": res["completed"] + 1,
```

Comment:
mypy is giving me `error: Value of type "Optional[Dict[str, Any]]" is not indexable  [index]` here and I have no idea what that means :upside_down_face: 

https://github.com/matrix-org/synapse/pull/10142/checks?check_run_id=2847878288#step:5:14

Reply:
Ah, because `res` is `Optional[Dict[str, Any]]`, which may be either `Dict[str, Any]` or `None`, and you can't do `None[...]`.

According to `simple_select_one_txn`s docstring, this is a known bug:

https://github.com/matrix-org/synapse/blob/1092718cac3800080bb766b251ae472282aef751/synapse/storage/database.py#L1711-L1714

So I would just override `res`s type to be `Dict[str, Any]` and leave a comment explaining why doing so is necessary.

```python
            res: Dict[str, Any] = self.db_pool.simple_select_one_txn(
                txn,
                "registration_tokens",
                keyvalues={"token": token},
                retcols=["pending", "completed"],
            )  # type: ignore
```

The `: Dict[str, Any]` overrides the type of `res`, and `type: ignore` tells mypy to ignore the otherwise dangerous type re-assignment.

Reply:
Ok, thanks.